rmpatch/rmpatch.c

```diff
@@ -9,20 +9,39 @@
 
 BOOL16 WINAPI InterruptRegister16(HTASK16 htask, FARPROC16 callback);
 BOOL16 WINAPI InterruptUnRegister16(HTASK16 htask);
+BOOL16 WINAPI NotifyRegister16(HTASK16 htask, FARPROC16 lpfnCallback, WORD wFlags);
+BOOL16 WINAPI NotifyUnRegister16(HTASK16 htask);
 BOOL16 WINAPI IsOldWindowsTask(HINSTANCE16 hinst);
 void WINAPI next_intcb(CONTEXT *context);
 
-void WINAPI checkpatch(NE_MODULE *pmodule, int segment, HANDLE16 *hseg)
+#include <pshpack1.h>
+typedef struct {
+	DWORD	dwSize;
+	WORD	wSelector;
+	WORD	wSegNum;
+	WORD	wType;		/* bit 0 set if this is a code segment */
+	WORD	wcInstance;	/* only valid for data segment */
+	SEGPTR  lpstrModuleName;
+} NFYLOADSEG;
+#include <poppack.h>
+
+BOOL WINAPI checkpatch(WORD id, DWORD data)
 {
-	char *mem = GlobalLock16(hseg);
-	char *name = (char *)pmodule + pmodule->ne_restab;
+	if (id != 1 /*NFY_LOADSEG*/)
+		return;
+	NFYLOADSEG *seg = (NFYLOADSEG *)MapSL(data);
+	char *name = (char *)MapSL(seg->lpstrModuleName);
+	if (GetExpWinVer16(GetModuleHandle16(name)) >= 0x300)
+		return;
+	char *mem = MapSL(MAKELONG(0, seg->wSelector));
+	WORD segment = seg->wSegNum;
 	if (!memcmp(mem, "MSEM87", 6)) // check for linked in MS 8087 emulator
 	{
 		FARPROC16 fpmath = GetProcAddress16(LoadLibrary16("WIN87EM.DLL"), "__FPMATH");
 		mem[0x24] = 0xea;  // JMP FAR
 		*(FARPROC16 *)(mem + 0x25) = fpmath;
 	}
-	else if (!strncmp(name + 1, "MGXWIN20", *name))
+	else if (!strncmp(name, "MGXWIN20", *name))
```

Comment:
invalid string length
```suggestion
	else if (!strcmp(name, "MGXWIN20"))
```
As with Windows 3.1, lpstrModuleName is specified as the first element of the resident names table, which is not null-terminated unless the low-order byte of the ordinal number are 0, but in this case it can be treated as null-terminated.